### DISCUSSION TITLE
`searchParams` on the Page context vs `useSearchParams`

### DISCUSSION THREAD

**Discussion by devjiwonchoi** [vercel/next.js#58351] [App Router]
IMO the `searchParams` that is being passed down to the Page component as the context is a successor to the `useSearchParams` hook.

What would be the use cases that `useSearchParams` are preferred over the context `searchParams`?

### Reference

- [`searchParams`](https://nextjs.org/docs/app/api-reference/file-conventions/page#searchparams-optional)
- [`useSearchparams`](https://nextjs.org/docs/app/api-reference/functions/use-search-params)

**Comment by laugharn** [comment]
searchParams is not available on static generation.

**Reply by devjiwonchoi** [reply]
Yeah, but my question is what are the use cases that `useSearchParams` are preferred over `searchParams`?

**Reply by eric-burel** [reply]
It will keep the page static while `searchParams` will trigger per-request server-side rendering.

**Comment by prkagrawal** [comment]
1. When we need to use search parameters in components or layouts,
  - pass searchParams prop
  - useSearchParams hook (rerenders with latest search parameters)

2. [Static rendering](https://nextjs.org/docs/app/api-reference/functions/use-search-params#static-rendering)

3. Access to [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) interface, `searchParams` only returns a simple object

**Comment by eric-burel** [comment]
@devjiwonchoi searchParams from page props will trigger dynamic (server-side, request time) rendering. 
useSearchParams will keep the page static, but will use client-side rendering.

You can't have server-side static rendering and searchParams without client-side rendering (you can have static + client-side though) because they are not part of "generateStaticParams", [sadly](https://twitter.com/ericbureltech/status/1724711996283572528).

**Reply by franky47** [reply]
@eric-burel side question: is there a way to access the page props searchParams in a deeply nested server component?

Since Context is not available on the server and ServerContexts don't have consumers, it seems like prop drilling is the only option. 🫤

**Reply by eric-burel** [reply]
@franky47 `cache` is the equivalent to client-side context, [check my article about that](https://prismic.io/blog/advanced-nextjs-server-context) (note: the naming "server context" is an unfortunate choice of mine, I wrote the article before this feature was really documented and serverContext in Next/React is actually smth else, meant to share data between server and client) or [this SO post](https://stackoverflow.com/questions/66106408/retrieve-data-server-side-and-save-in-context-with-next-js/75533592#75533592).

Edit: also I've crafted an open source demo [here](https://github.com/eric-burel/next13-server-context/blob/main/src/app/getServerContext.ts)

Code would look like this:
```ts
// ctx.ts
import { cache } from "react";

export const getServerCache: () => { preview: boolean } = cache(() => ({
    preview: false
}))
...
// In your page
const ctx = getServerCache()
ctx.preview = !!searchParams.get("preview")
...
// in a component
const ctx = getServerCache()
if (ctx.preview) ...
```

Note: you might want to give it a shot I haven't used this pattern ever since I don't have this level of nesting.

**Reply by franky47** [reply]
I figured the cache was the way to go, but having not slept much last night, the solution wasn't immediate, thanks for this pointer! It didn't occur to me to mutate the cached object, but since only the object ref is cached, it makes sense.

I'm trying to provide an automated abstraction for next-usequerystate to feed it all the searchParams, have them parsed at the page top level, and access them as needed in server components.

Making it all type safe is going to be fun! 😊

**Comment by elijaholmos** [comment]
I'm curious if any of the above answers have changed with Next.js 15 and the necessity to invoke `use(searchParams)` in client-side components.

**Reply by wellitongervickas** [reply]
Not even the AI could help here hahaha

**Reply by wellitongervickas** [reply]
I'd suggest take a look here: https://github.com/vercel/next.js/discussions/58351#discussioncomment-7575734
